File: kotlin/FakeWorld/test/FakeWorldTest.kt
import warier.Elf
import warier.Fighter
import warier.Hobbit
import warier.Human
import warier.Wizard

/**
 * Test all classes working correctly or not.
 */
class FakeWorldTest {
    private lateinit var hobbit1: Hobbit
    private lateinit var hobbit2: Hobbit
    private lateinit var elf1: Elf
    private lateinit var elf2: Elf
    private lateinit var human1: Human
    private lateinit var human2: Human
    private lateinit var fighter1: Fighter
    private lateinit var fighter2: Fighter
    private lateinit var wizard1: Wizard
    private lateinit var wizard2: Wizard

    @org.junit.jupiter.api.BeforeEach
    fun setUp() {
        println("Initializing Test Warier...")

        hobbit1 = Hobbit("Frodo")
        hobbit2 = Hobbit("Samwise", 8, 12, 2, 14, 10, 30)

        elf1 = Elf("Legolas")
        elf2 = Elf("Elrond", 14, 20, 18, 12, 1000, 80, "Forest", hobbit1)

        human1 = Human("Aragorn")
        human2 = Human("Wormtongue", 100, 12, 2, 16, 100, 50, elf2)

        fighter1 = Fighter("Boromir")
        fighter2 = Fighter("Faramir", 18, 15, 3, 8, 150, 55, elf1)

        wizard1 = Wizard("Gandalf")
        wizard2 = Wizard("Saruman", 8, 12, 4, 15, 2500, 90, elf1, 20)

        println("\nTwo Hobbits arrived:")
        println(hobbit1)
        println(hobbit2)

        println("\nTwo Elves arrived:")
        println(elf1)
        println(elf2)

        println("\nTwo Humans arrived:")
        println(human1)
        println(human2)

        println("\nTwo Fighters arrived:")
        println(fighter1)
        println(fighter2)

        println("\nTwo Wizards arrived:")
        println(wizard1)
        println(wizard2)
    }

    @org.junit.jupiter.api.AfterEach
    fun tearDown() = println("Test War Completed...")

    @org.junit.jupiter.api.Test
    fun test() {
        println("Starting Test War...")

        System.out.printf("\n%s attacks %s! \n", elf2.name, human2.name)
        elf2.attack(human2)
        println(human2)

        System.out.printf("\n%s steals from %s!\n", hobbit1.name, human2.name)
        hobbit1.stealFrom(human2)
        println(hobbit1)
        println(human2)

        System.out.printf("\n%s changes his best friend!\n", elf2.name)
        elf2.becomeFriendOf(hobbit2)
        println(elf2)

        System.out.printf("\n%s attacks %s!\n", fighter1.name, hobbit1.name)
        fighter1.attack(hobbit1)
        println(hobbit1)

        System.out.printf("\n%s heals %s!\n", wizard1.name, hobbit1.name)
        wizard1.heal(hobbit1)
        println(wizard1)
        println(hobbit1)

        System.out.printf("\n%s attacks itself!\n", fighter1.name)
        println(fighter1)
        fighter1.attack(fighter1)
        println(fighter1)
    }
}

File: flutter/fake_world/test/fake_world_test.dart
import 'package:fake_world/warrior/elf.dart';
import 'package:fake_world/warrior/fighter.dart';
import 'package:fake_world/warrior/hobbit.dart';
import 'package:fake_world/warrior/human.dart';
import 'package:fake_world/warrior/humanoid.dart';
import 'package:fake_world/warrior/wizard.dart';
import 'package:test/test.dart';

void main() {
  final world = FakeWorldTest();
  world.setUp();
  test('FakeWorld complete test...', () => world.test());
  world.tearDown();
}

class FakeWorldTest {
  late final Hobbit hobbit1;
  late final Hobbit hobbit2;
  late final Elf elf1;
  late final Elf elf2;
  late final Human human1;
  late final Human human2;
  late final Fighter fighter1;
  late final Fighter fighter2;
  late final Wizard wizard1;
  late final Wizard wizard2;

  void setUp() {
    print('Initializing test warriors...\n');

    hobbit1 = Hobbit(name: 'Frodo');
    hobbit2 = Hobbit(
      name: 'Samwise',
      dexterity: 8,
      strength: 12,
      armour: 2,
      moxie: 14,
      coins: 10,
      health: 30,
    );

    elf1 = Elf(name: 'Legolas');
    elf2 = Elf(
      name: 'Elrond',
      dexterity: 14,
      strength: 20,
      armour: 18,
      moxie: 12,
      coins: 1000,
      health: 80,
      clan: 'Forest',
      friend: hobbit1,
    );

    human1 = Human(name: 'Aragorn');
    human2 = Human(
      name: 'Wormtongue',
      dexterity: 100,
      strength: 12,
      armour: 2,
      moxie: 16,
      coins: 100,
      health: 50,
      enemy: elf2,
    );

    fighter1 = Fighter(name: 'Boromir');
    fighter2 = Fighter(
      name: 'Faramir',
      dexterity: 18,
      strength: 15,
      armour: 3,
      moxie: 8,
      coins: 150,
      health: 55,
      enemy: elf1,
    );

    wizard1 = Wizard(name: 'Gandalf');
    wizard2 = Wizard(
      name: 'Saruman',
      dexterity: 8,
      strength: 12,
      armour: 4,
      moxie: 15,
      coins: 2500,
      health: 90,
      enemy: elf1,
      magicRating: 20,
    );

    print('\nTwo Hobbits arrived:\n');
    print(hobbit1);
    print(hobbit2);

    print('\nTwo Elves arrived:\n');
    print(elf1);
    print(elf2);

    print('\nTwo Humans arrived:\n');
    print(human1);
    print(human2);

    print('\nTwo Fighters arrived:\n');
    print(fighter1);
    print(fighter2);

    print('\nTwo Wizards arrived:\n');
    print(wizard1);
    print(wizard2);
  }

  void test() {
    print('Starting test war...');

    print('\n${elf2.name} attacks ${human2.name}! \n');
    elf2.attack(human2);
    print(human2);

    print('\n${hobbit1.name} steals from ${human2.name}!\n');
    hobbit1.stealFrom(human2);
    print(hobbit1);
    print(human2);

    print('\n${elf2.name} changes his best friend!\n');
    elf2.becomeFriendOf(hobbit2);
    print(elf2);

    print('\n${fighter1.name} attacks ${hobbit1.name}!\n');
    fighter1.attack(hobbit1);
    print(hobbit1);

    print('\n${wizard1.name} heals ${hobbit1.name}!\n');
    wizard1.heal(hobbit1);
    print(wizard1);
    print(hobbit1);

    print('\n${fighter1.name} attacks itself!\n');
    print(fighter1);
    fighter1.attack(fighter1);
    print(fighter1);
  }

  void tearDown() => print('\nTest war completed...');
}
